frame/utility/src/lib.rs

```diff
@@ -497,6 +497,39 @@ pub mod pallet {
 			let res = call.dispatch_bypass_filter(frame_system::RawOrigin::Root.into());
 			res.map(|_| ()).map_err(|e| e.error)
 		}
+
+		/// Dispatches a function call with a provided origin. Unlike `dispatch_as`, this
+		/// function propagates errors returned by the wrapped `call`.
+		///
+		/// The dispatch origin for this call must be _Root_.
+		///
+		/// # <weight>
+		/// - O(1).
+		/// - Limited storage reads.
+		/// - One DB write (event).
+		/// - Weight of derivative `call` execution + T::WeightInfo::dispatch_as().
+		/// # </weight>
+		#[pallet::call_index(6)]
+		#[pallet::weight({
+			let dispatch_info = call.get_dispatch_info();
+			(
+				T::WeightInfo::dispatch_as()
+					.saturating_add(dispatch_info.weight),
+				dispatch_info.class,
+			)
+		})]
+		pub fn ensure_dispatch_as(
+			origin: OriginFor<T>,
+			as_origin: Box<T::PalletsOrigin>,
+			call: Box<<T as Config>::RuntimeCall>,
+		) -> DispatchResult {
+			ensure_root(origin)?;
+
+			call.dispatch_bypass_filter((*as_origin).into()).map_err(|e| e.error)?;
+
+			Self::deposit_event(Event::DispatchedAs { result: Ok(()) });
```

Comment:
The result is not always `Ok`.

Reply:
If the result is not `Ok`, then the function returns early because of the `?` after the `dispatch_bypass_filter`. Or do you mean something else?

Reply:
Hm, yea but then no event is emitted. The `dispatch_as` always emits an event also in the error case. Not sure how to keep that semantic here, or if it even makes sense.

Reply:
only success extrinsic should emit event so this is expected behaviour